### ISSUE TITLE
Chain over the "rejected" side of an `Async`

### ISSUE THREAD

**Issue by mgreystone** [evilsoft/crocks#393]
I would like to request the ability to chain over the left or "rejected" side of an `Async`. This would be helpful to move values from the left side of the `Async` to the right.

```js
getAsync()
  .chainRej(ifElse(isErrorOk, constant(Async.of('Error Ignored')), Async.Rejected))
```

I'm currently doing this with a helper function.
```js
// chainRej :: (a -> Async e a) -> Async a b -> Async e a|b
export const chainRej = curry((fn, a) =>
  Async((rej, res) => a.fork(x => fn(x).fork(rej, res), res))
)
```

**Comment by dalefrancis88** [comment]
Hey @mgreystone have you checked out [`Coalesce`](https://evilsoft.github.io/crocks/docs/crocks/Async.html#coalesce) or [`Swap`](https://evilsoft.github.io/crocks/docs/crocks/Async.html#swap) I think that they may be more what you're looking for? Coalesce takes two functions one for the left and one for the right to map the values to the same codomain and swap moves left into right and right into left

**Comment by dalefrancis88** [comment]
Product types will always only `map`, `chain` etc over the 'right' value except where the function itself is designed to consider both parts i.e. `coalesce` and `swap. There is also another common method called `bimap` similiar to `swap` but each function could have a different co-domain and they each remain in there respective sides

**Comment by mgreystone** [comment]
Hi @dalefrancis88 . I don't think `coalesce`, `swap`, `map`, `chain` or `bimap` will do what i'm looking for.
* `coalesce` will always move the value from left-to-right, whereas i'm looking to do this conditionally.
* `swap` moves the left to the right & the right to to the left. I'm afraid i'm not sure how i could conditionally move from left-to-right with `swap`.
* `map` only affects the right side
* `chain` is the reverse of what i want. I can move a value from right-to-left.
* `bimap` does not exchange values between sides at all.

**Comment by evilsoft** [comment]
I like it. Think we should figure this one out.
we could add it as a special method on the type instance and then provide an Async specific pointfree function. Would like to see this in on the next patch release. we will be having `0.12.0` coming soon, and we will follow it up in `0.12.1`.
Unless @dalefrancis88 or @karthikiyengar or @bennypowers sees any issue with that. But nothing comes to mind that could be a show stopper for this VERY helpful function.

Then as people request it for other Sum types, we can address, but I would say starting with `Async` will get the best value.

**Comment by dalefrancis88** [comment]
What about the idea of doing this as an implementation of `bichain`? If we did that it could open for more use cases and have a nice symmetry with `map`/`bimap`?

**Comment by evilsoft** [comment]
ohhh @dalefrancis88 that sounds great.
@mgreystone  What do you think about that...This would mean you would do the same pattern as we do with using `bimap`. Except if you ONLY want to work the `Rejected` side, you will have to provide `Async.of` for the `Resolve` side (instead of `identity` like we would do with `bimap`).

But if we go that route, we could implement it for every `Sum` type, except for `Maybe`. Then again `Maybe` might be cool for a in-type `option`, but the fact we would have to have the Nothing portion ALWAYS be `Unit`. So that means, if for some reason you passed `a -> b`, it will always be called as `() -> b`, which could be problematic. So maybe we start with just everything but `Maybe`.

**Comment by mgreystone** [comment]
`bichain` sounds like a win to me.

**Comment by evilsoft** [comment]
Sweet, if you are good with that @dalefrancis88 then will get this in soon yeah?

**Comment by dalefrancis88** [comment]
Yeah, i'll get started on it today

**Comment by evilsoft** [comment]
@mgreystone Sorry it is taking so long, has to get that 0.12.0 release done, This is now the next thing on the list, and we'll cut a release once it is all in there.

**Comment by dalefrancis88** [comment]
I should find some time this week to work on this. It's close to done

**Comment by evilsoft** [comment]
@mgreystone we are still pounding this out. Should be 🔴 👁 soon. Will cut a release when this is in master.
We are going to add this to many of the other Sum Types, but we will get Async out first.

**Comment by dalefrancis88** [comment]
@mgreystone There is a PR ready for `bichain` sitting here, https://github.com/evilsoft/crocks/pull/394 if you have time, check it out and let me know if you see any opportunities for improvement

**Comment by evilsoft** [comment]
This has shipped with `0.12.3`. Closing!
